File: 0001-0500/0438. Find All Anagrams in a String.java
// Medium
// Hash Table, Sliding Window
// Time:O(m + n), Space: O(k)
// https://leetcode.cn/problems/find-all-anagrams-in-a-string/

import java.util.*;

class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        Map<Character, Integer> need = new HashMap<>();
        Map<Character, Integer> window = new HashMap<>();
        for (char c : p.toCharArray()) {
            need.put(c, need.getOrDefault(c, 0) + 1);
        }

        List<Integer> res = new ArrayList<>();
        int left = 0, right = 0, valid = 0;
        while (right < s.length()) {
            // expand window
            char c = s.charAt(right);
            right++;

            if (need.containsKey(c)) {
                window.put(c, window.getOrDefault(c, 0) + 1);
                if (window.get(c).equals(need.get(c))) {
                    valid++;
                }
            }

            if (right - left >= p.length()) {
                if (valid == need.size())
                    res.add(left);
                char drop = s.charAt(left);
                left++;
                if (need.containsKey(drop)) {
                    if (window.get(drop).equals(need.get(drop))) {
                        valid--;
                    }
                    window.put(drop, window.get(drop) - 1);
                }
            }
        }
        return res;
    }
}

File: 0001-0500/0438. Find All Anagrams in a String.go
package main
// Medium
// Time:O(n), Space:O(1)
// https://leetcode.cn/problems/find-all-anagrams-in-a-string/

func findAnagrams(s string, p string) []int {
    res := []int{}
    if len(s) < len(p) {
        return res
    }

    // count p's freq
    pCount := make(map[byte]int)
    for i := 0; i < len(p); i++ {
        pCount[p[i]]++
    }

    windowCount := make(map[byte]int)
    for i := 0; i < len(s); i++ {
        windowCount[s[i]]++

        if i >= len(p) {
            left := s[i - len(p)]
            windowCount[left]--
            if windowCount[left] == 0 {
                delete(windowCount, left)
            }
        }
        
        if i >= len(p) - 1 && checkEqual(pCount, windowCount) {
            res = append(res, i - len(p) + 1)
        }
    }
    return res
}

func checkEqual(m1, m2 map[byte]int) bool {
    if len(m1) != len(m2) {
        return false
    }
    for k, v := range m1 {
        if m2[k] != v {
            return false
        }
    }
    return true
}
